engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -71,16 +71,36 @@ public interface IR {
    */
   List<IR> children();
 
+  /**
+   * Applies the callback to nodes in the preorder walk of the tree of this node.
+   *
+   * @param ir the node to traverse
+   * @param cb the callback to apply
+   */
+  static void preorder(IR ir, Consumer<IR> cb) {
+    final class CB implements scala.Function1<IR, scala.Unit> {
+      @Override
+      public scala.Unit apply(IR ir) {
+        cb.accept(ir);
+        ir.children().foreach(this);
+        return null;
+      }
+    }
+
+    cb.accept(ir);
+    ir.children().foreach(new CB());
+  }
+
   /**
    * Lists all the nodes in the preorder walk of the tree of this node.
    *
    * @return all the descendants of this node.
    */
   default List<IR> preorder() {
```

Comment:
This method could also be turned into `static` utility method. Keeping it as `default` instance method is _"only good"_ for compatibility with previous code. In any case this method shall never be overridden.

Reply:
Btw. do we need this method still? I see many passes using the new callback approach. Where do we still use this _build whole list_ approach?

Reply:
It is used mostly in tests. And there are two usages left: in `AliasAnalysis` and `DataflowAnalysis` compiler passes because we traverse two IR trees side by side and still need some queue there